xds/internal/resolver/xds_resolver_test.go

```diff
@@ -1106,54 +1110,137 @@ func (s) TestResolverWRR(t *testing.T) {
+	hcm := testutils.MarshalAny(&v3httppb.HttpConnectionManager{
+		RouteSpecifier: &v3httppb.HttpConnectionManager_Rds{Rds: &v3httppb.Rds{
+			ConfigSource: &v3corepb.ConfigSource{
+				ConfigSourceSpecifier: &v3corepb.ConfigSource_Ads{Ads: &v3corepb.AggregatedConfigSource{}},
 			},
+			RouteConfigName: rdsName,
+		}},
+		HttpFilters: []*v3httppb.HttpFilter{e2e.RouterHTTPFilter},
+		CommonHttpProtocolOptions: &v3corepb.HttpProtocolOptions{
+			MaxStreamDuration: durationpb.New(1 * time.Second),
 		},
-	}, nil)
+	})
+	resources := e2e.UpdateOptions{
+		NodeID: nodeID,
+		Listeners: []*v3listenerpb.Listener{{
+			Name:        ldsName,
+			ApiListener: &v3listenerpb.ApiListener{ApiListener: hcm},
+			FilterChains: []*v3listenerpb.FilterChain{{
+				Name: "filter-chain-name",
+				Filters: []*v3listenerpb.Filter{{
+					Name:       wellknown.HTTPConnectionManager,
+					ConfigType: &v3listenerpb.Filter_TypedConfig{TypedConfig: hcm},
+				}},
+			}},
+		}},
+		Routes: []*v3routepb.RouteConfiguration{{
+			Name: rdsName,
+			VirtualHosts: []*v3routepb.VirtualHost{{
+				Domains: []string{ldsName},
+				Routes: []*v3routepb.Route{
+					{
+						Match: &v3routepb.RouteMatch{PathSpecifier: &v3routepb.RouteMatch_Prefix{Prefix: "/foo"}},
+						Action: &v3routepb.Route_Route{Route: &v3routepb.RouteAction{
+							ClusterSpecifier: &v3routepb.RouteAction_WeightedClusters{WeightedClusters: &v3routepb.WeightedCluster{
+								Clusters: []*v3routepb.WeightedCluster_ClusterWeight{
+									{
+										Name:   "A",
+										Weight: &wrapperspb.UInt32Value{Value: 100},
+									},
+								}},
+							},
+							MaxStreamDuration: &v3routepb.RouteAction_MaxStreamDuration{
+								MaxStreamDuration: durationpb.New(5 * time.Second),
+							},
+						}},
+					},
+					{
+						Match: &v3routepb.RouteMatch{PathSpecifier: &v3routepb.RouteMatch_Prefix{Prefix: "/bar"}},
+						Action: &v3routepb.Route_Route{Route: &v3routepb.RouteAction{
+							ClusterSpecifier: &v3routepb.RouteAction_WeightedClusters{WeightedClusters: &v3routepb.WeightedCluster{
+								Clusters: []*v3routepb.WeightedCluster_ClusterWeight{
+									{
+										Name:   "B",
+										Weight: &wrapperspb.UInt32Value{Value: 100},
+									},
+								}},
+							},
+							MaxStreamDuration: &v3routepb.RouteAction_MaxStreamDuration{
+								MaxStreamDuration: durationpb.New(0 * time.Second),
+							},
+						}},
+					},
+					{
+						Match: &v3routepb.RouteMatch{PathSpecifier: &v3routepb.RouteMatch_Prefix{Prefix: "/"}},
+						Action: &v3routepb.Route_Route{Route: &v3routepb.RouteAction{
+							ClusterSpecifier: &v3routepb.RouteAction_WeightedClusters{WeightedClusters: &v3routepb.WeightedCluster{
+								Clusters: []*v3routepb.WeightedCluster_ClusterWeight{
+									{
+										Name:   "C",
+										Weight: &wrapperspb.UInt32Value{Value: 100},
+									},
+								}},
+							},
+						}},
+					},
+				},
+			}},
+		}},
+		SkipValidation: true,
+	}
```

Comment:
This is a really long inline update sitting in the test body. Is there any reusability of this with other inline updates elsewhere in this file? Can we pull this into a helper (I'm thinking of e2e_tests and DefaultXXXResource() with some wrappers for knobs)?

Reply:
Thanks for the explanation. If there's a bunch of default boilerplate as part of the proto (that gets normal e2e flow), I feel like it saves a ton of lines to have a general helper and either a. another helper that wraps that helper which sets a field (in this case max stream duration) or call the helper and modify the resource in the test. I wasn't reading it regardless, the comment was in relation to a lot of extra lines in the test.